### ISSUE TITLE
[ERROR] Could not resolve "os"

### ISSUE THREAD

**Issue by kaiyoma** [evanw/esbuild#2409]
I'm trying to build a command-line tool that runs Cypress (the e2e testing framework), which in turn loads a plugin that imports the `os` module, but I keep getting this error at runtime:

```
X [ERROR] Could not resolve "os"

    ../../common/temp/node_modules/.pnpm/cypress-fail-on-console-error@3.2.0/node_modules/cypress-fail-on-console-error/dist/index.js:32:19:
      32 │ var os_1 = require("os");
         ╵                    ~~~~

  The package "os" wasn't found on the file system but is built into node. Are you trying to bundle for node? You can use "platform: 'node'" to do that, which will remove this error.
```

However, I'm already specifying `platform: "node"`, as you can see in my `esbuild.config.js`:

```
const esbuild = require('esbuild');

const packageJson = require('./package.json');

esbuild
  .build({
    bundle: true,
    entryPoints: ['src/index.ts'],
    external: Object.keys(packageJson.dependencies),
    format: 'cjs',
    minify: true,
    outdir: 'dist',
    platform: 'node',
    target: 'node16',
  })
  .catch(() => process.exit(1));
```

I've already tried adding `os` to `external` and that didn't help. What am I doing wrong here?

Node: 16.13.0
esbuild: 0.14.38

**Comment by evanw** [comment]
I'm not sure. Either adding `os` to external or setting the platform to `node` should work.

Please provide a code sample that would let me reproduce the issue.

**Comment by kaiyoma** [comment]
I've gotten a little bit further here, but still having issues. I'll upload a repro in a bit.

If I try adding `external: ['os']` to the esbuild options, I get this error:

```
> Dynamic require of "os" is not supported
```

If I try adding `platform: 'node'` to the esbuild options, I get this error:

```
> require is not defined
```

**Comment by kaiyoma** [comment]
Here's the best approximation I can cobble together: [esbuild-repro.zip](https://github.com/evanw/esbuild/files/9193751/esbuild-repro.zip)

Unzip and run: `npm i && npx cypress run cypress/e2e/test.cy.js`

My project is actually a command-line tool, built with esbuild, that wraps around `cypress run`, which uses esbuild to build test code, so it's a little more complicated, but I'm seeing the same kinds of errors in this simple repro too.

**Comment by evanw** [comment]
Sorry but it doesn't run (complaining about missing dependencies), and it looks like installing Cypress is pretty involved. Can you provide a way to reproduce this that just involves esbuild? You said you're getting the error with `esbuild.build()`. Could you provide a script to run `esbuild.build()` and the problematic inputs?

**Comment by kaiyoma** [comment]
Hmm, installing Cypress is actually pretty easy, so I'm not sure why it's not working for you. `npm i` works just fine for me:

```
$ npm i

added 191 packages, and audited 192 packages in 11s

28 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
```

What error are you getting?

The problem I'm running into is with `cypress-fail-on-console-error` specifically so Cypress is unfortunately necessary here. BTW, I found a related issue, but that also seems to be open without a real solution: https://github.com/evanw/esbuild/issues/1921

**Comment by hyrious** [comment]
This should be a problem in `cypress-fail-on-console-error` which imports node builtin modules in the browser. Setting `platform: node` or mark `os` as external is wrong here. What we have to do is make polyfills to make that module work in the browser.

**Comment by kaiyoma** [comment]
AFAICT The plugin doesn't run in the browser. The plugin is running in Node and I see this error on the command line, before the browser is launched.

**Comment by hyrious** [comment]
No, this plugin runs in the browser as [Cypress Doc says](https://docs.cypress.io/guides/references/trade-offs#Inside-the-browser). The plugin works by hooking the `console.error` in browser context and sending the message to the node.js side so that it can print error in your command line. This plugin is very short so you can just check its source code: https://unpkg.com/browse/cypress-fail-on-console-error@3.2.1/dist/index.js as you can see it does import a node module `var os_1 = require("os");` and it does communicate with cypress with `Cypress.log`.

I guess other people are not facing this problem because their bundlers already do polyfills for them, like webpack4 or webpack5 with some plugins, or [parcel2](https://parceljs.org/features/node-emulation/#polyfilling-%26-excluding-builtin-node-modules).

**Comment by kaiyoma** [comment]
True, test code runs in the browser, but plugins are different, as you can read [here](https://docs.cypress.io/guides/tooling/plugins-guide):

> Normally, as a user, all of your test code, your application, and Cypress commands are executed in the browser. But Cypress is also a Node process that plugins can use. Plugins enable you to tap into the Node process running outside of the browser.

**Comment by kaiyoma** [comment]
I messed around with this some more today and figured out how to get Node module polyfills working in esbuild, and that seems to have done the trick. For posterity, this is the plugin I added:

https://github.com/remorses/esbuild-plugins#esbuild-pluginsnode-modules-polyfill

Just adding that (without changing any `platform` or `external` settings) was enough to get things working again. Thanks for the suggestion @hyrious!
